### PR TITLE
statistics: shard needsStatsMap 

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: ref #51853

Problem Summary:

### What changed and how does it work?

there is a global lock, and although there are plenty of write/read paths, by sharding, I can increase the granularity of the locks.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/statistics/table.go [hunk 1]
```diff
@@ -668,24 +668,73 @@ func (n *neededStatsMap) AllItems() []model.TableItemID {
...
 }
 
+const shardCnt = 128
+
+type neededStatsMap struct {
+	items [shardCnt]neededStatsInternalMap
+}
+
+func getIdx(tbl model.TableItemID) int64 {
+	result := tbl.ID % shardCnt
```
Comment (AilinKid) [inline pkg/statistics/table.go:696]: you can just tbl.ID's negativeness before mod

Comment (hawkingrei) [reply pkg/statistics/table.go]: Done

**Review by winoros — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Pull request merged**